presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PrefilterForLimitingAggregation.java

```diff
@@ -0,0 +1,266 @@
+            TypeProvider types,
+            PlanVariableAllocator variableAllocator,
+            PlanNodeIdAllocator idAllocator,
+            WarningCollector warningCollector)
+    {
+        if (SystemSessionProperties.isPrefilterForGroupbyLimit(session)) {
+            return SimplePlanRewriter.rewriteWith(new Rewriter(session, metadata, types, idAllocator, variableAllocator), plan);
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final TypeProvider types;
+        private final PlanNodeIdAllocator idAllocator;
+        private final PlanVariableAllocator variableAllocator;
+        private Rewriter(
+                Session session,
+                Metadata metadata,
+                TypeProvider types,
+                PlanNodeIdAllocator idAllocator,
+                PlanVariableAllocator variableAllocator)
+        {
+            this.session = session;
+            this.metadata = metadata;
+            this.types = types;
+            this.idAllocator = idAllocator;
+            this.variableAllocator = variableAllocator;
+        }
+
+        @Override
+        public PlanNode visitSort(SortNode sortNode, RewriteContext<Void> context)
+        {
+            return sortNode;
+        }
+
+        @Override
+        public PlanNode visitLimit(LimitNode limitNode, RewriteContext<Void> context)
+        {
+            PlanNode source = limitNode.getSource();
+            AggregationNode aggregationNode;
+
+            if (source instanceof ProjectNode && ((ProjectNode) source).getSource() instanceof AggregationNode) {
+                aggregationNode = (AggregationNode) ((ProjectNode) source).getSource();
+            }
+            else if (source instanceof AggregationNode) {
+                aggregationNode = (AggregationNode) source;
+            }
+            else {
+                return limitNode;
+            }
+
+            if (!aggregationNode.getGroupingKeys().isEmpty() && isScanFilterProject(aggregationNode.getSource())) {
+                PlanNode rewrittenAggregation = addPrefilter(aggregationNode, limitNode.getCount());
+                if (rewrittenAggregation == aggregationNode) {
+                    return limitNode;
+                }
+
+                PlanNode newLimitNode;
+                if (source == aggregationNode) {
+                    newLimitNode = replaceChildren(limitNode, ImmutableList.of(rewrittenAggregation));
+                }
+                else {
+                    newLimitNode = replaceChildren(limitNode, ImmutableList.of(replaceChildren(source, ImmutableList.of(rewrittenAggregation))));
+                }
+
+                return newLimitNode;
+            }
+
+            return limitNode;
+        }
+
+        private PlanNode addPrefilter(AggregationNode aggregationNode, long count)
```

Comment:
%s/count/limit ?

Reply:
It's the count not limit - we need to exactly load those many. The intent